File: magpie/src/pixel_1dto3d.f90

subroutine pix_id_1dto3d_scalar(xpix_id, ypix_id, zpix_id, ygrid, zgrid, pix_id)

  ! Returns the combined 3d grid pixel from pixel indices across each axis.
  !
  ! Parameters
  ! ----------
  ! xpix_id : int
  !   Pixel indices along the x axis grid.
  ! ypix_id : int
  !   Pixel indices along the y axis grid.
  ! zpix_id : int
  !   Pixel indices along the z axis grid.
  ! ygrid : int
  !   Length of y axis grid.
  ! zgrid : int
  !   Length of z axis grid.
  !
  ! Returns
  ! -------
  ! pix_id : int
  !   3d pixel index.

  implicit none

  integer, intent(in) :: ygrid, zgrid
  integer, intent(in) :: xpix_id, ypix_id, zpix_id
  integer, intent(out) :: pix_id

  ! Main

  if ((xpix_id .NE. -1) .AND. (ypix_id .NE. -1) .AND. (zpix_id .NE. -1)) then
    pix_id = zpix_id + zgrid*(ypix_id + ygrid*xpix_id)
  else
    pix_id = -1
  end if

end subroutine pix_id_1dto3d_scalar


subroutine pix_id_3dto1d_scalar(pix_id, ygrid, zgrid, xpix_id, ypix_id, zpix_id)

  ! Returns the combined 3d grid pixel from pixel indices across each axis.
  !
  ! Parameters
  ! ----------
  ! pix_id : int
  !   3d pixel index.
  ! ygrid : int
  !   Length of y axis grid.
  ! zgrid : int
  !   Length of z axis grid.
  !
  ! Returns
  ! -------
  ! xpix_id : int
  !   Pixel indices along the x axis grid.
  ! ypix_id : int
  !   Pixel indices along the y axis grid.
  ! zpix_id : int
  !   Pixel indices along the z axis grid.

  implicit none

  integer, intent(in) :: ygrid, zgrid
  integer, intent(in) :: pix_id
  integer, intent(out) :: xpix_id, ypix_id, zpix_id
  integer :: xypix_id

  ! Main

  if (pix_id .EQ. -1) then
    xpix_id = -1
    ypix_id = -1
    zpix_id = -1
    xypix_id = 0
  else
    xypix_id = int(floor(real(pix_id/zgrid)))
    zpix_id = pix_id - zgrid*xypix_id
    xpix_id = int(floor(real(xypix_id/ygrid)))
    ypix_id = xypix_id - ygrid*xpix_id
  end if

end subroutine pix_id_3dto1d_scalar


subroutine pix_id_1dto3d_grid(xpix_id, ypix_id, zpix_id, xlen, ylen, zlen &
  , ygrid, zgrid, pix_id)

  ! Maps pixels given along a single axis in x, y and z onto a 3d grid
  ! flattened.
  !
  ! Parameters
  ! ----------
  ! xpix_id : int
  !   Pixel indices along the x axis grid.
  ! ypix_id : int
  !   Pixel indices along the y axis grid.
  ! zpix_id : int
  !   Pixel indices along the z axis grid.
  ! xlen : int
  !   Size of the xpix_id.
  ! ylen : int
  !   Size of the ypix_id.
  ! zlen : int
  !   Size of the zpix_id.
  ! ygrid : int
  !   Length of y axis grid.
  ! zgrid : int
  !   Length of z axis grid.
  !
  ! Returns
  ! -------
  ! pix_id : int
  !   3d grid pixel.

  implicit none

  integer, intent(in) :: xlen, ylen, zlen, ygrid, zgrid
  integer, intent(in) :: xpix_id(xlen), ypix_id(ylen), zpix_id(zlen)
  integer, intent(out) :: pix_id(xlen*ylen*zlen)

  integer :: i, j, k, ii

  ! Main

  ii = 1

  do i = 1, xlen
    do j = 1, ylen
      do k = 1, zlen
        call pix_id_1dto3d_scalar(xpix_id(i), ypix_id(j), zpix_id(k), ygrid, zgrid, pix_id(ii))
        ii = ii + 1
      end do
    end do
  end do

end subroutine pix_id_1dto3d_grid


subroutine pix_id_1dto3d_array(xpix_id, ypix_id, zpix_id, xlen, ygrid, zgrid, pix_id)

  ! Returns the combined grid pixel from pixel indices across each axis.
  !
  ! Parameters
  ! ----------
  ! xpix_id : int
  !   Pixel indices along the x axis grid.
  ! ypix_id : int
  !   Pixel indices along the y axis grid.
  ! zpix_id : int
  !   Pixel indices along the z axis grid.
  ! xlen : int
  !   Size of the xpix_id, ypix_id and zpix_id.
  ! ygrid : int
  !   Length of y axis grid.
  ! zgrid : int
  !   Length of z axis grid.
  !
  ! Returns
  ! -------
  ! pix_id : int
  !   3d array pixel.

  implicit none

  integer, intent(in) :: xlen, ygrid, zgrid
  integer, intent(in) :: xpix_id(xlen), ypix_id(xlen), zpix_id(xlen)
  integer, intent(out) :: pix_id(xlen)

  integer :: i

  ! Main

  do i = 1, xlen
    call pix_id_1dto3d_scalar(xpix_id(i), ypix_id(i), zpix_id(i), ygrid, zgrid, pix_id(i))
  end do

end subroutine pix_id_1dto3d_array



subroutine pix_id_3dto1d_array(pix_id, xlen, ygrid, zgrid, xpix_id, ypix_id, zpix_id)

  ! Returns pixel indices across each axis from the combined grid pixel.
  !
  ! Parameters
  ! ----------
  ! pix_id : int
  !   3d array pixel.
  ! xlen : int
  !   Size of the xpix_id, ypix_id and zpix_id.
  ! ygrid : int
  !   Length of y axis grid.
  ! zgrid : int
  !   Length of z axis grid.
  !
  ! Returns
  ! -------
  ! xpix_id : int
  !   Pixel indices along the x axis grid.
  ! ypix_id : int
  !   Pixel indices along the y axis grid.
  ! zpix_id : int
  !   Pixel indices along the z axis grid.

  implicit none

  integer, intent(in) :: xlen, ygrid, zgrid
  integer, intent(in) :: pix_id(xlen)
  integer, intent(out) :: xpix_id(xlen), ypix_id(xlen), zpix_id(xlen)

  integer :: i

  ! Main

  do i = 1, xlen
    call pix_id_3dto1d_scalar(pix_id(i), ygrid, zgrid, xpix_id(i), ypix_id(i), zpix_id(i))
  end do

end subroutine pix_id_3dto1d_array

File: magpie/srcpy/pixel_1dto3d.py
import numpy as np
from numba import njit


@njit
def pix_id_1dto3d_scalar(xpix_id, ypix_id, zpix_id, ygrid, zgrid):
    """
    Converts 3D pixel indices into a single 1D index.

    Parameters
    ----------
    xpix_id : int
        Pixel index along the x-axis grid.
    ypix_id : int
        Pixel index along the y-axis grid.
    zpix_id : int
        Pixel index along the z-axis grid.
    ygrid : int
        Length of y-axis grid.
    zgrid : int
        Length of z-axis grid.

    Returns
    -------
    pix_id : int
        3D pixel index.
    """
    if xpix_id == -1 or ypix_id == -1 or zpix_id == -1:
        return -1
    else:
        return zpix_id + zgrid * (ypix_id + ygrid * xpix_id)


@njit
def pix_id_3dto1d_scalar(pix_id, ygrid, zgrid):
    """
    Converts a single 1D pixel index back into 3D pixel indices.

    Parameters
    ----------
    pix_id : int
        3D pixel index.
    ygrid : int
        Length of y-axis grid.
    zgrid : int
        Length of z-axis grid.

    Returns
    -------
    xpix_id : int
        Pixel index along the x-axis grid.
    ypix_id : int
        Pixel index along the y-axis grid.
    zpix_id : int
        Pixel index along the z-axis grid.
    """
    if pix_id == -1:
        return -1, -1, -1
    else:
        xypix_id = pix_id // zgrid
        zpix_id = int(pix_id - zgrid*xypix_id)
        xpix_id = xypix_id // ygrid
        ypix_id = int(xypix_id - ygrid*xpix_id)
        return xpix_id, ypix_id, zpix_id


@njit
def pix_id_1dto3d_grid(xpix_id, ypix_id, zpix_id, ygrid, zgrid):
    """
    Maps 3D grid pixel indices to a 1D flattened grid.

    Parameters
    ----------
    xpix_id : array-like
        Pixel indices along the x-axis grid.
    ypix_id : array-like
        Pixel indices along the y-axis grid.
    zpix_id : array-like
        Pixel indices along the z-axis grid.
    ygrid : int
        Length of y-axis grid.
    zgrid : int
        Length of z-axis grid.

    Returns
    -------
    pix_id : array-like
        Flattened 3D grid pixel indices.
    """
    xlen, ylen, zlen = len(xpix_id), len(ypix_id), len(zpix_id)
    pix_id = np.empty(xlen * ylen * zlen, dtype=np.int32)

    ii = 0
    for i in range(xlen):
        for j in range(ylen):
            for k in range(zlen):
                pix_id[ii] = pix_id_1dto3d_scalar(xpix_id[i], ypix_id[j], zpix_id[k], ygrid, zgrid)
                ii += 1

    return pix_id


@njit
def pix_id_1dto3d_array(xpix_id, ypix_id, zpix_id, ygrid, zgrid):
    """
    Converts arrays of 3D pixel indices into a 1D pixel index array.

    Parameters
    ----------
    xpix_id : array-like
        Pixel indices along the x-axis grid.
    ypix_id : array-like
        Pixel indices along the y-axis grid.
    zpix_id : array-like
        Pixel indices along the z-axis grid.
    ygrid : int
        Length of y-axis grid.
    zgrid : int
        Length of z-axis grid.

    Returns
    -------
    pix_id : array-like
        Flattened 3D pixel indices.
    """
    xlen = len(xpix_id)
    pix_id = np.empty(xlen, dtype=np.int32)

    for i in range(xlen):
        pix_id[i] = pix_id_1dto3d_scalar(xpix_id[i], ypix_id[i], zpix_id[i], ygrid, zgrid)

    return pix_id


@njit
def pix_id_3dto1d_array(pix_id, ygrid, zgrid):
    """
    Converts an array of 1D pixel indices into separate x, y, and z pixel index arrays.

    Parameters
    ----------
    pix_id : array-like
        Flattened 3D pixel indices.
    ygrid : int
        Length of y-axis grid.
    zgrid : int
        Length of z-axis grid.

    Returns
    -------
    xpix_id : array-like
        Pixel indices along the x-axis grid.
    ypix_id : array-like
        Pixel indices along the y-axis grid.
    zpix_id : array-like
        Pixel indices along the z-axis grid.
    """
    xlen = len(pix_id)
    xpix_id = np.empty(xlen, dtype=np.int32)
    ypix_id = np.empty(xlen, dtype=np.int32)
    zpix_id = np.empty(xlen, dtype=np.int32)

    for i in range(xlen):
        xpix_id[i], ypix_id[i], zpix_id[i] = pix_id_3dto1d_scalar(pix_id[i], ygrid, zgrid)

    return xpix_id, ypix_id, zpix_id
